<?xml version="1.0" encoding="utf-8"?>

<launch>

    <!-- Load params and args -->
    <arg 
        name="model_name" 
        default="flat_model_circles" />
    <arg 
        name="param_file" 
        default="/config/video_9976.yaml" />
    <rosparam 
        command="load" 
        file="$(find sofa_valid)$(arg param_file)" />

    <!-- Start Sofa Simulation -->
    <!-- ========================================================================== -->
    <param 
        name="sofa_scene_path" 
        type="string" 
        value="$(find sofa_valid)/scenes/valid_exp.pyscn"/>
    <param 
        name="sofa_binary_path" 
        type="string" 
        value="/home/roland/sofa/build/bin/runSofa"/>
    <param 
        name="sofa_gui_type" 
        type="string" 
        value="qglviewer"/>
    <node 
        pkg="sofa_sim" 
        type="sofa_run_node.py" 
        name="sofa_valid" 
        output="screen">
    </node>

    <!-- Load parameters -->


    <!-- Launch Navion -->
    <!-- ================================ -->
    <arg 
        name="cal_path" 
        default="$(find mpem)/cal/Navion_2_Calibration_24-02-2020.yaml" />
    <arg 
        name="emns_param_path" 
        default="$(find mag_manip)/params/Navion_2.yaml" />
    
    <node
        name="backward_model" 
        pkg="mag_calculator" 
        type="backward_model" 
        output="screen" 
        respawn="true">
            <param 
                name="calibration_path" 
                type="string" 
                value="$(arg cal_path)" />
            <param 
                name="is_field_inverted" 
                type="bool" 
                value="true" />
            <param 
                name="max_field_intensity" 
                type="double" 
                value="0.08" />
            <remap 
                from="/backward_model/currents" 
                to="navion/target_currents"/>
    </node>

    <node 
        name="forward_model" 
        pkg="mag_calculator" 
        type="forward_model" 
        output="screen" 
        respawn="true">
            <param 
                name="calibration_path" 
                type="string" 
                value="$(arg cal_path)" />
            <param 
                name="display_warning_zero_field" 
                type="bool" 
                value="false" />
            <param 
                name="is_field_inverted" 
                type="bool" 
                value="true" />
            <remap 
                from="/forward_model/currents" 
                to="navion/actual_currents"/>
            <remap 
                from="/forward_model/field_array" 
                to="navion/actual_field"/>
    </node>
<!-- 
    <node 
        pkg="tsc_utils" 
        type="poses_custom_grid_node.py" 
        name="forward_model_list_pub" 
        respawn="true">
            <param name="~frame_id" type="string" value="mns" />
            <param name="~nx" type="int" value="7" />
            <param name="~ny" type="int" value="5" />
            <param name="~nz" type="int" value="1" />
            <param name="~spacing" type="double" value="0.05" />
            <remap from="~/poses" to="/forward_model/poses"/>
    </node>  -->

    <!-- Mesh models -->
    <include 
        file="$(find nav_rviz)/launch/navion_urdf_model.launch"/>

    <include file="$(find sofa_valid)/launch/flat_model_urdf.launch">
       <arg
          name="model" 
          value="$(find sofa_valid)/urdf/$(arg model_name).urdf"/>
    </include>

    <!-- Show video in rviz main viewer
    <include 
        file="$(find sofa_valid)/launch/textured_quad_video.launch"/> -->


    <!-- rviz -->
    <!-- ================================ -->
    <node 
        pkg="rviz" 
        name="rviz" 
        type='rviz' 
        args="-d $(find sofa_valid)/rviz/sofa_sim.rviz" />
        

</launch>

